xdsl/dialects/hw.py

```diff
@@ -112,6 +121,142 @@ def print_parameters(self, printer: Printer) -> None:
         printer.print_string(">")
 
 
+class InnerSymbolTable(OpTrait):
+    """A trait for inner symbol table functionality on an operation.
+    Merges the upstream table of inner symbols and their resolutions and the op trait.
+    """
+
+    @overload
+    def __init__(self) -> None:
+        # Create an empty table, not used for lookups
+        ...
+
+    @overload
+    def __init__(self, op: Operation) -> None:
+        ...
+
+    def __init__(self, op: Operation | None = None) -> None:
+        self.symbol_table: Mapping[StringAttr, InnerSymTarget] = dict()
```

Comment:
Can you please move the property declaration to the class level? I'm surprised we don't have a ruff rule for this, but we've converged on pretty much always declaring the full list of properties before the inits.